src/mono/wasm/runtime/binding_support.js

```diff
@@ -143,32 +143,58 @@ var BindingSupportLib = {
 		mono_array_to_js_array: function (mono_array) {
 			if (mono_array == 0)
 				return null;
+			
+			var array_root = MONO.mono_wasm_new_root (mono_array), 
+				elem_root = MONO.mono_wasm_new_root ();
 
 			var res = [];
 			var len = this.mono_array_length (mono_array);
 			for (var i = 0; i < len; ++i)
 			{
-				var ele = this.mono_array_get (mono_array, i);
-				if (this.is_nested_array(ele))
-					res.push(this.mono_array_to_js_array(ele));
+				let ele = elem_root.value = this.mono_array_get (mono_array, i);
+				if (this.is_nested_array (ele))
+					res.push (this.mono_array_to_js_array(ele));
 				else
 					res.push (this.unbox_mono_obj (ele));
 			}
 
+			array_root.release();
+			elem_root.release();
+
 			return res;
 		},
 
 		js_array_to_mono_array: function (js_array) {
-			var mono_array = this.mono_obj_array_new (js_array.length);
+			var array_root = MONO.mono_wasm_new_root (this.mono_obj_array_new (js_array.length)),
+				elem_root = MONO.mono_wasm_new_root ();
+				
 			for (var i = 0; i < js_array.length; ++i) {
-				this.mono_obj_array_set (mono_array, i, this.js_to_mono_obj (js_array [i]));
+				elem_root.value = this.js_to_mono_obj (js_array [i]);
+				this.mono_obj_array_set (array_root.value, i, elem_root.value);
 			}
+
+			array_root.release();
+			elem_root.release();
+
 			return mono_array;
 		},
 
+		// Scratch root used for temporary storage in unbox_mono_obj.
+		// As long as this API is not re-entrant, this single slot is sufficient to ensure
+		//  that the object being unboxed is not collected until the unbox operation is done.
+		_unbox_mono_obj_root: null,
+
+		// FIXME: Is there a way to mechanically ensure that this function is not re-entrant,
+		//  other than the overhead of a try/finally block?
 		unbox_mono_obj: function (mono_obj) {
 			if (mono_obj == 0)
 				return undefined;
+
+			if (!MONO._unbox_mono_obj_root)
+				MONO._unbox_mono_obj_root = MONO.mono_wasm_new_root ();
+
+			MONO._unbox_mono_obj_root.value = mono_obj;
+			
```

Comment:
This needs to be cleared at the end.

Reply:
yeah I think i'm just gonna try/finally all the functions that use roots as convention so it's fool-proof